core/App/ehsm_provider.h

```diff
@@ -29,120 +29,74 @@
-
-#define EH_AES_GCM_IV_SIZE  12
-#define EH_AES_GCM_MAC_SIZE 16
-
-#define RSA_OAEP_2048_SHA_256_MAX_ENCRYPTION_SIZE       190
-//#define RSA_2048_OAEP_SHA_1_MAX_ENCRYPTION_SIZE       214
-
-#define RSA_OAEP_3072_SHA_256_MAX_ENCRYPTION_SIZE       318
-//#define RSA_3072_OAEP_SHA_1_MAX_ENCRYPTION_SIZE       342
-
-#define SM2PKE_MAX_ENCRYPTION_SIZE                      6047
-
-#define RSA_OAEP_3072_CIPHER_LENGTH       384
-#define RSA_OAEP_3072_SIGNATURE_SIZE      384
-#define RSA_OAEP_3072_DIGEST_SIZE         256
-
-typedef enum {
-    EH_OK                           = 0,
-    EH_KEYSPEC_INVALID              = -1,
-    EH_DEVICE_MEMORY                = -2,
-    EH_DEVICE_ERROR                 = -3,
-    EH_GENERAL_ERROR                = -4,
-    EH_FUNCTION_FAILED              = -5,
-    EH_ARGUMENTS_BAD                = -6,
-    EH_LA_SETUP_ERROR               = -7,
-    EH_LA_EXCHANGE_MSG_ERROR        = -8,
-    EH_LA_CLOSE_ERROR               = -9,
-} ehsm_status_t;
-
-typedef enum  {
-    EH_INTERNAL_KEY,
-    EXTERNAL_KEY,
-} ehsm_keyorigin_t;
-
-typedef enum {
-    EH_AES_GCM_128 = 0,
-    EH_AES_GCM_256 =1,
-    EH_RSA_2048 = 2,
-    EH_RSA_3072 = 3,
-    EH_EC_P256 = 4,
-    EH_EC_P512 = 5,
-    EH_EC_SM2 =6,
-    EH_SM4 =7,
-} ehsm_keyspec_t;
-
-typedef enum {
-    ENCRYPT_DECRYPT = 0,
-    SIGN_VERIFY = 1,
-} ehsm_keypurpose_t;
-
-#pragma pack(push,1)
-
-typedef struct {
-    uint32_t keyspec;
-    uint32_t origin;
-    uint32_t purpose;
-    uint32_t apiversion;
-    uint8_t  descrption[16];
-    uint8_t  createdate[8];
-} ehsm_keymetadata_t;
-
-typedef struct {
-    ehsm_keymetadata_t  metadata;
-    uint32_t            keybloblen;
-    uint8_t             *keyblob;
-} ehsm_keyblob_t;
-
-typedef struct {
-    uint32_t    datalen;
-    uint8_t     *data;
-} ehsm_data_t;
-
-#pragma pack(pop)
-
-#ifndef SAFE_FREE
-#define SAFE_FREE(ptr) {if (NULL != (ptr)) {free(ptr); (ptr) = NULL;}}
-#endif
-
-
-namespace EHsmProvider
+typedef enum
+{
+    EH_INITIALIZE = 0,
+    EH_FINALIZE,
+    EH_CREATE_KEY,
+    EH_ENCRYPT,
+    EH_DECRYPT,
+    EH_ASYMMETRIC_ENCRYPT,
+    EH_ASYMMETRIC_DECRYPT,
+    EH_SIGN,
+    EH_VERIFY,
+    EH_GENERATE_DATAKEY,
+    EH_GENERATE_DATAKEY_WITHOUT_PLAINTEXT,
+    EH_EXPORT_DATAKEY,
+    EH_GET_VERSION,
+    EH_ENROLL,
+    EH_GENERATE_QUOTE,
+    EH_VERIFY_QUOTE,
+} ehsm_action_t;
+
+extern "C"
 {
 
+    /**
+     * @brief The unique ffi entry for the ehsm provider libaray.
+     *
+     * @param paramJson the request parameters in the form of JSON string
+     * [string] json string
+        {
+            Action: string [CreateKey, Encrypt, Decrypt, Sign, Verify...]
+            payload: {
+                [additional parameter]
+            }
+        }
+     *
+     * @return char* response in json string
+        {
+            code: int,
+            message: string,
+            result: {
+                xxx : xxx
+            }
+        }
+     */
+    char *EHSM_NAPI_CALL(const char *paramJson);
```

Comment:
rename it to EHSM_FFI_CALL, if the sub function is called as ffi_xxx

Reply:
It will be change